<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Running cNMTF</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #0000ff; } /* Keyword */
code > span.ch { color: #008080; } /* Char */
code > span.st { color: #008080; } /* String */
code > span.co { color: #008000; } /* Comment */
code > span.ot { color: #ff4000; } /* Other */
code > span.al { color: #ff0000; } /* Alert */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #008000; font-weight: bold; } /* Warning */
code > span.cn { } /* Constant */
code > span.sc { color: #008080; } /* SpecialChar */
code > span.vs { color: #008080; } /* VerbatimString */
code > span.ss { color: #008080; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #0000ff; } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ff4000; } /* Preprocessor */
code > span.do { color: #008000; } /* Documentation */
code > span.an { color: #008000; } /* Annotation */
code > span.cv { color: #008000; } /* CommentVar */
code > span.at { } /* Attribute */
code > span.in { color: #008000; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="preprocessing.html">Your input data</a>
</li>
<li>
  <a href="running.html">Running cNMTF</a>
</li>
<li>
  <a href="prioritising.html">Prioritising SNVs</a>
</li>
<li>
  <a href="extras.html">Extras</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Running cNMTF</h1>

</div>


<style type="text/css">

body{
  font-size: 13pt !important;;
}
</style>
<hr />
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The <span class="math inline">\(R\)</span> matrix is approximated as the product of low-dimensional matrices <span class="math inline">\(U\)</span>, <span class="math inline">\(V\)</span> and <span class="math inline">\(S\)</span>. Here, we use <span class="math inline">\(L_U\)</span>, <span class="math inline">\(V_o\)</span> and <span class="math inline">\(A\)</span> to penalize the factorisation and guide the solutions of <span class="math inline">\(U\)</span> and <span class="math inline">\(V\)</span>.</p>
<p><img src="images/factorisation.png" width="700px"></p>
<p>This dimensional reduction provides information for clustering tasks, so <span class="math inline">\(U\)</span> and <span class="math inline">\(V\)</span> are taken as cluster indicator matrices for SNVs and patients respectively. Simultaneously, we compute the product of <span class="math inline">\(U\)</span> and <span class="math inline">\(S\)</span> to generate a score matrix <span class="math inline">\(\Omega\)</span>. This matrix summarizes the effect of single SNVs on clusters of patients with specific outcomes and can be used to prioritise the SNVs.</p>
<p><br></p>
<hr />
</div>
<div id="consensus-clustering-and-snv-score" class="section level2">
<h2>Consensus clustering and SNV score</h2>
<p><br></p>
<p>You can run cNMTF with one single function: <code>score.cnmtf</code>. This function reads your input data, finds the optimal parameters for the algorithm and provides you with the clusters of patients and the SNV scores, <span class="math inline">\(\Omega\)</span>.</p>
<p><br></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&#39;doParallel&#39;</span>)
<span class="kw">library</span>(<span class="st">&quot;igraph&quot;</span>)
<span class="kw">library</span>(<span class="st">&#39;cnmtf&#39;</span>)

    <span class="kw">score.cnmtf</span>( <span class="dt">R =</span> R, <span class="co">#Relationship matrix</span>
                 <span class="dt">out =</span> out, <span class="co">#Categorical outcome variable</span>
                 <span class="dt">pop =</span> pop, <span class="co">#Population variable</span>
                 <span class="dt">log.file =</span> <span class="st">&quot;logfile_my_experiment&quot;</span>, <span class="co">#Log-file to track the performance</span>

                 <span class="co">#Variables to Save/Load data and workspaces</span>
                 <span class="dt">name.exp =</span> <span class="st">&quot;my_experiment&quot;</span>, <span class="co">#Name of experiment to save files</span>
                 <span class="dt">work.dat =</span> <span class="st">&quot;./test/&quot;</span>, <span class="co">#Folder to save and load workspaces</span>

                 <span class="co">#Number of clusters</span>
                 <span class="dt">define.k =</span> <span class="st">&quot;method&quot;</span>, <span class="co">#Method to define k1</span>
                 <span class="dt">k1 =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">15</span>,<span class="dv">20</span>), <span class="co">#Number of clusters of SNVs</span>
                 <span class="dt">k2 =</span> <span class="kw">nlevels</span>(out), <span class="co">#Number of clusters of patients</span>

                 <span class="co">#Penalisation parameters</span>
                 <span class="dt">wparameters =</span> <span class="kw">list</span>( <span class="dt">gamma1 =</span> <span class="fl">0.5</span>, <span class="co">#Weight for the SNV network, Lu</span>
                                     <span class="dt">gamma2 =</span> <span class="dv">1</span>, <span class="co">#Weight for the outcome matrix, Vo</span>
                                     <span class="dt">gamma3 =</span> <span class="dv">1</span>), <span class="co">#Weight for the kernel matrix, A</span>
                 <span class="dt">save.parameters =</span> <span class="ot">TRUE</span>, <span class="co">#Save parameters to file</span>
                 <span class="dt">run.t.par =</span> <span class="dv">4</span>, <span class="co">#Number of repetitions for parameters fitting</span>
                 <span class="dt">max.try0 =</span> <span class="dv">4</span>,  <span class="co">#Maximum number of tries to fit the parameter</span>
                 <span class="dt">snps.known =</span> <span class="ot">NULL</span>, <span class="co">#List of known SNV associated with the trait</span>

                 <span class="co">#Variables to control performance of the algorithm</span>
                 <span class="dt">parallel.opt =</span> <span class="ot">TRUE</span>, <span class="co">#Run some instances of the algorithm in parallel</span>
                 <span class="dt">n.cores =</span> <span class="dv">3</span>, <span class="co">#Number of cores to use in the parallel processing</span>
                 <span class="dt">init =</span> <span class="dv">0</span>, <span class="co">#Type of seeding/initialisation of matrices in the algorithm</span>
                 <span class="dt">do.U =</span> <span class="ot">TRUE</span>, <span class="co">#Perform clustering of SNPs</span>
                 <span class="dt">calcObj =</span> <span class="dv">20</span>, <span class="co">#Check convergency every 20 iterations</span>
                 <span class="dt">calcObj2 =</span> <span class="dv">40</span>, <span class="co">#Start checking convergency after first 40 iterations</span>
                 <span class="dt">iters =</span> <span class="dv">300</span>, <span class="co">#Number of iterations</span>
                 <span class="dt">run.t.exp =</span> <span class="dv">10</span>, <span class="co">#Number of repetitions for the experiment</span>
                 <span class="dt">display.iters =</span> <span class="ot">FALSE</span>, <span class="co">#Display the iterations of function cnmtf</span>

                 <span class="co">#Randomisations</span>
                 <span class="dt">score.pvalues =</span> <span class="ot">TRUE</span>, <span class="co">#Estimate p-values for the scores</span>
                 <span class="dt">random.parallel =</span> <span class="ot">TRUE</span>, <span class="co">#Run each randomisation in parallel</span>
                 <span class="dt">randomisations =</span> <span class="dv">100</span>, <span class="co">#Number of randomisations</span>

                 <span class="co">#Construction of penalisation terms</span>
                 <span class="dt">file.Gu =</span> <span class="st">&quot;./test/Gu_ppi_test.RData&quot;</span> <span class="co">#Workspace with SNV-SNV network</span>

            )</code></pre></div>
<p><strong>The arguments for this function are explained as follows.</strong></p>
<p><br></p>
<hr />
<div id="the-input-data" class="section level3">
<h3>The input data</h3>
<p>You only must provide the relationship matrix, <code>R</code>, the categorical outcome variable, <code>out</code>, the population origin of patients if known, <code>pop</code> and the SNV-SNV network’s workspace, <code>file.Gu</code> (generated with the function <code>create.network</code>). The algorithm will call internally the functions to create the outcome matrix <span class="math inline">\(V_o\)</span> and the kernel of ancestry, <span class="math inline">\(A\)</span>.</p>
<p><br></p>
<hr />
</div>
<div id="the-output-data" class="section level3">
<h3>The output data</h3>
<p>The output files will be labelled by the name of the experiment, <code>name.exp</code>, and redirected to the folder that you specifiy in <code>work.dat</code>. The algorithm will generate a <code>log.file</code> which reports every internal procedure and might be very useful to catch problems in your data.</p>
<p><br></p>
<hr />
</div>
<div id="define-the-number-of-clusters" class="section level3">
<h3>Define the number of clusters</h3>
<p>By default, the number of clusters of patients equals the number of levels in the outcome (<em>i.e.,</em> 2 clusters for case-control designs). On the other hand, the algorithm will try to find the optimal number of clusters for the SNVs. This is based on a grid search (<span class="math inline">\(k_1: \{2, 3, 5, 10, \dots \}\)</span> ) while tracking the cluster stability. We use a dispersion coefficient, <span class="math inline">\(\rho_{k_1}\)</span>, to summarize the consistency of clustering assignments throughout repetitions of the algorithm. You must define the values of the grid search.</p>
<center>
<p><img src="images/k1k2_my_experiment.pdf" width="400px"></p>
<p><strong>Figure. Finding optimal <span class="math inline">\(k_1\)</span>.</strong> The <em>dispersion coefficient</em> is tracked for a grid of <span class="math inline">\(k_1\)</span>. values. The <em>dispersion coefficient</em> reaches a plateau around <span class="math inline">\(k_1 = 10\)</span>, which is the optimum value selected.</p>
</center>
<p><br></p>
<hr />
</div>
<div id="estimate-the-optimal-penalisation-parameters" class="section level3">
<h3>Estimate the optimal penalisation parameters</h3>
<p>You must weight how much information is transferred from the data sources (<span class="math inline">\(L_u\)</span>, <span class="math inline">\(A\)</span> and <span class="math inline">\(V_o\)</span>) into <span class="math inline">\(U\)</span> and <span class="math inline">\(V\)</span>. By default, the network and genotyping data contribute equally in the final SNV clusters (<span class="math inline">\(w_{L_U} = 0.5\)</span>). Also, we aim at achieving maximum separation of cases-controls while minimising the effect of population structures embedded in <span class="math inline">\(R\)</span>; therefore, the contribution from <span class="math inline">\(A\)</span> and <span class="math inline">\(V_o\)</span> is maximum (<span class="math inline">\(w_{V_o} = 1\)</span> and <span class="math inline">\(w_{A} = 1\)</span>). The algorithm will read a list of weights <code>list(gamma1 = 0.5, gamma2 = 1, gamma3 = 1)</code> specifying the contribution from <span class="math inline">\(L_u\)</span>, <span class="math inline">\(V_o\)</span> and <span class="math inline">\(A\)</span>, respectively.</p>
<p>cNMTF will estimate the optimal parameters that fit the weights provided above. Each parameter is estimated individually in the following order: <span class="math inline">\(\gamma_2\)</span>, <span class="math inline">\(\gamma_3\)</span> and <span class="math inline">\(\gamma_1\)</span>. For each parameter, cNMTF will create a grid search of 4 elements, <code>max.try0</code> (<em>i.e</em>, number of tries to fit the parameter). At each of this tries the algorithm will be repeated 4 times, <code>run.t.par</code>, and consensus clustering results generated. These consensus clusters are tracked with two variables:</p>
<ul>
<li>We track how the <em>total node degree</em> changes when <span class="math inline">\(\gamma_1\)</span> increase.</li>
<li>We track the similarity between clusters, outcome and population ancestry of the patients when <span class="math inline">\(\gamma_2\)</span> and <span class="math inline">\(\gamma_3\)</span> change.</li>
</ul>
<center>
<p><img src="images/parameter_gamma1_my_experiment.pdf" width="300px"> <img src="images/parameter_gamma2_my_experiment.pdf" width="300px"> <img src="images/parameter_gamma3_my_experiment.pdf" width="300px"></p>
<p><strong>Figure. Finding optimal penalisation parameters.</strong> The <em>total node degree</em> is tracked for a grid of <span class="math inline">\(\gamma_1\)</span> values. The <em>similarity cluster-outcome</em> and <em>similarity cluster-population</em> are tracked for a grid of <span class="math inline">\(\gamma_2\)</span> and <span class="math inline">\(\gamma_3\)</span> values, respectively. Optimal parameters are: <span class="math inline">\(\gamma_1 = 12\)</span>, <span class="math inline">\(\gamma_2 = 8\)</span> and <span class="math inline">\(\gamma_3 = 0.25\)</span>.</p>
</center>
<p>See more details about the tracking variables in our publication.</p>
<p>Once the algorithm finds such penalisation parameters, everything is ready to conduct the factorisation and generate the final results.</p>
<p><br></p>
<hr />
</div>
<div id="controlling-the-performance-of-the-algorithm" class="section level3">
<h3>Controlling the performance of the algorithm</h3>
<p>The algorithm can run some instances in parallel depending on the available RAM memory of your machine (<code>parallel.opt = TRUE</code>). By default, it will run 10 repetitions of the algorithm (<code>run.t.exp = 10</code>), through the processor cores that you make available (<code>n.cores = 3</code>). For example, the first 3 repetitions of cNMTF will be processed in parallel, after them, another batch of 3 repetitions will be forked in the 3 cores of your machine. The process continues till completing all 10 repetitions. Each core will run a factorisation of your big matrices so requiring enough RAM memory in your machine. It is advisable to keep the number of cores low if your are running cNMTF locally in your regular desktop.</p>
<p>At each repetition, the algorithm will minimise an objective function. This procedure requires some parameters too:</p>
<ul>
<li><p>Initialisation of <span class="math inline">\(U\)</span>, <span class="math inline">\(S\)</span> and <span class="math inline">\(V\)</span> matrices. It can be done randomly, <code>init = 0</code>, or by following a Single Value Decomposition based approach, <code>init = 1</code> (Boutsidis <em>et al.</em>, 2008).</p></li>
<li>The algorithm performs iterations to find the optimal <span class="math inline">\(U\)</span>, <span class="math inline">\(S\)</span> and <span class="math inline">\(V\)</span> matrices. To stop the iterations, any of the following convergence criteria must be fulfilled:
<ul>
<li>The maximum number of iterations is reached (<code>iters = 300</code>).</li>
<li>The minimum relative change in the objective function is reached (<code>tof = 1e-05</code>).</li>
</ul>
<p>Such criteria are evaluated after the firt first 40 iterations (<code>calcObj2</code>) and then every 20 iterations (<code>calcObj</code>).</p></li>
</ul>
<p><br></p>
<hr />
</div>
<div id="creating-randomised-results" class="section level3">
<h3>Creating randomised results</h3>
<p>After finishing the main clustering step, the algorithm has the option of factorising randomisations of <span class="math inline">\(R\)</span>, by using the same optimal parameters. We will use these randomisations later to estimate p-values for the SNV scores. You must define whether the randomisations are conducted with <code>score.pvalues = T</code>, the total number : <code>randomisations = 100</code> and whether they are computed in parallel (<code>random.parallel = TRUE</code>).</p>
<hr />
<p>To this point, the algorithm has processed your data and directed its outputs to “./test/”. Your workspace is now ready for <a href="prioritising.html">prioritising variants</a>.</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
